### PR TITLE
Pickup ASAN and LLVM assertion build fixes from #43418

### PR DESCRIPTION
**Pull request by vchuravy**
- fix compiler warning in cli
- fix LLVM_ASSERTIONS build
- sanitizers: fix support loading of libjulia-codegen
- sanitizers: disable calls to pthread_atfork

Cherry-picked from #43418

### CODE REVIEW

**Comment by DilumAluthge** [comment]
If you like, you can also push a revert of https://github.com/JuliaLang/julia/pull/43663 to this branch.

**Comment by vchuravy** [comment]
> If you like, you can also push a revert of #43663 to this branch.

The real question is whether or not this unbreaks ASAN.

**Comment by DilumAluthge** [comment]
We'll find out in a few minutes 😂

**Comment by vchuravy** [comment]
Looks very promising!

**Comment by DilumAluthge** [comment]
`asan` is green: https://buildkite.com/julialang/julia-master/builds/7258#_

**Comment by DilumAluthge** [comment]
`llvmpasses` is still red: https://buildkite.com/julialang/julia-master/builds/7258#9c7688cd-61a6-40fc-a949-71f6a3971664

File: cli/loader_lib.c [hunk 1]
```diff
@@ -198,6 +198,11 @@ __attribute__((constructor)) void jl_load_libjulia_internal(void) {
 
     // Unpack our special library names.  This is why ordering of library names matters.
     libjulia_internal = load_library(special_library_names[0], lib_dir, 1);
+
+    void (*pZN4llvm2cl22ResetCommandLineParserEv)(void) = (void (*)(void))lookup_symbol(libjulia_internal, "_ZN4llvm2cl22ResetCommandLineParserEv");
+    if (pZN4llvm2cl22ResetCommandLineParserEv)
+       pZN4llvm2cl22ResetCommandLineParserEv();
```
Comment (tkf) [inline cli/loader_lib.c:204]: This is rather gross so I wonder if it makes sense to put it behind `ifdef`? It at least clarifies when it's needed.

```suggestion
#if defined(_COMPILER_ASAN_ENABLED_) || defined(_COMPILER_TSAN_ENABLED_)
    void (*pZN4llvm2cl22ResetCommandLineParserEv)(void) = (void (*)(void))lookup_symbol(libjulia_internal, "_ZN4llvm2cl22ResetCommandLineParserEv");
    if (pZN4llvm2cl22ResetCommandLineParserEv)
       pZN4llvm2cl22ResetCommandLineParserEv();
#endif
```

**Review by tkf — COMMENTED** [review]
The C++ mangling is terrifying but anything that makes ASAN work LGTM :)

File: cli/loader_lib.c (hunk 1 above)
Comment (vchuravy) [reply cli/loader_lib.c:204]: @vtjnash thoughts?

Comment (vtjnash) [reply cli/loader_lib.c:204]: For some reason, I had thought the need for this had went away with LLVM13. Is that not true? My testing might have been flawed.

Comment (vchuravy) [reply cli/loader_lib.c]: Let's see. I pushed a revert and CI should tell us if we still need it.

Comment (vtjnash) [reply cli/loader_lib.c]: Looks like it sometimes works now

**Comment by vchuravy** [comment]
But something is very wrong, I triggered another buiild before the revert commit https://buildkite.com/julialang/julia-master/builds/7542 IIRC asan was working okay then instead of having `NaN` everywhere...

**Pull request merged**